### PR TITLE
ConditionalNodes representing WDL if blocks

### PR DESCRIPTION
**Pull request by cjllanwarne**

### CODE REVIEW

File: wom/src/main/scala/wdl4s/wdl/If.scala [hunk 1]
```diff
@@ -29,3 +23,28 @@ case class If(index: Int, condition: WdlExpression, ast: Ast)
...
+    */
+  def apply(ast: Ast, index: Int): If = {
+    new If(index, WdlExpression(ast.getAttribute("expression")), ast)
+  }
+
+  def womConditionalNode(ifBlock: If, localLookup: Map[String, GraphNodePort.OutputPort]): ErrorOr[ConditionalNodeWithInputs] = {
+    val ifConditionExpression = WdlWomExpression(ifBlock.condition, Option(ifBlock))
+    val ifConditionGraphInputExpressionValidation = WdlWomExpression.findInputsforExpression("conditional", ifConditionExpression, localLookup, Map.empty)
+    val ifConditionTypeValidation = ifConditionExpression.evaluateType(localLookup.map { case (k, v) => k -> v.womType }) flatMap {
+      case WdlBooleanType => Valid(()) // Covers maps because this is a custom unapply
```
Comment (mcovarr) [inline wom/src/main/scala/wdl4s/wdl/If.scala]: I do not understand these words here

File: wom/src/main/scala/wdl4s/wdl/If.scala [hunk 2]
```diff
@@ -29,3 +23,28 @@ case class If(index: Int, condition: WdlExpression, ast: Ast)
...
+      case other => s"Cannot scatter over a non-array type ${other.toWdlString}".invalidNel
```
Comment (mcovarr) [inline wom/src/main/scala/wdl4s/wdl/If.scala]: oh this is just copy paste fail from scatter code

File: wom/src/main/scala/wdl4s/wom/graph/ConditionalNode.scala [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+import wdl4s.wdl.types.{WdlBooleanType, WdlOptionalType}
+import wdl4s.wom.graph.GraphNode.LinkedInputPort
+
+/**
+  * Currently only WDL has the concept of conditional executions:
+  *
+  * @param innerGraph Imagine that the contents of the conditional block were a self-contained workflow. That's this Graph
+  * @param condition The (boolean) expression on which the conditional is predicated.
+  * @param innerGraphInputs The set of input ports that will be needed to satisfy the inner graph's inputs.
+  * @param outputMapping Output ports for the scatter node, which also link back to GraphOutputNodes of the inner graph.
```
Comment (mcovarr) [inline wom/src/main/scala/wdl4s/wom/graph/ConditionalNode.scala]: excellent docs!

File: wom/src/main/scala/wdl4s/wom/graph/ConditionalNode.scala [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+  final case class ConditionalNodeWithInputs(conditionalNode: ConditionalNode, inputs: Set[GraphInputNode])
+
+  def wireInConditional(innerGraph: Graph, condition: GraphNodeInputExpression, inputMapping: Map[String, OutputPort]): ErrorOr[ConditionalNodeWithInputs] = {
+    val graphNodeSetter = new GraphNode.GraphNodeSetter()
+    val inputPortLinker = GraphNode.linkInputPort("", inputMapping, graphNodeSetter.get) _
+
+    val conditionalExpressionType: ErrorOr[Unit] = condition.evaluateType flatMap {
+      case WdlBooleanType => Valid(())
+      case other => s"Cannot base a conditional on non-boolean type ${other.toWdlString}".invalidNel
+    }
```
Comment (mcovarr) [inline wom/src/main/scala/wdl4s/wom/graph/ConditionalNode.scala]: This whole block looks nearly identical to the one above, is there a DRYing opportunity?

File: wom/src/main/scala/wdl4s/wdl/If.scala [hunk 5]
```diff
@@ -29,3 +23,28 @@ case class If(index: Int, condition: WdlExpression, ast: Ast)
...
+    }
+
+    val innerGraphValidation: ErrorOr[Graph] = WdlGraphNode.buildWomGraph(ifBlock, Set.empty, localLookup)
+
+    (ifConditionGraphInputExpressionValidation, ifConditionTypeValidation, innerGraphValidation) flatMapN { (ifConditionGraphInputExpression, conditionType, innerGraph) =>
```
Comment (mcovarr) [inline wom/src/main/scala/wdl4s/wdl/If.scala]: underscore the unused type validation

File: wom/src/main/scala/wdl4s/wom/graph/GraphNodePort.scala [hunk 6]
```diff
@@ -41,10 +41,13 @@ object GraphNodePort {
...
 
-  // TODO: For these next two, the graphNode should be a ScatterNode and IfNode respectively (once those exist):
   /**
     * Represents the gathered output from a call/declaration in a scatter.
     */
   final case class ScatterGathererPort(name: String, womType: WdlArrayType, outputToGather: PortBasedGraphOutputNode, g: Unit => GraphNode) extends OutputPort with DelayedGraphNodePort
-  final case class ConditionalOutputPort(name: String, womType: WdlOptionalType, outputToExpose: OutputPort, graphNode: GraphNode) extends OutputPort
+
+  /**
+    * Represents the conditional output from a
```
Comment (mcovarr) [inline wom/src/main/scala/wdl4s/wom/graph/GraphNodePort.scala]: looks like you had something insightful to say but then

File: wom/src/main/scala/wdl4s/wdl/WdlGraphNode.scala [hunk 7]
```diff
@@ -85,10 +86,21 @@ object WdlGraphNode {
...
+
+      case ifBlock: If => If.womConditionalNode(ifBlock, acc.availableInputs) map { case ConditionalNodeWithInputs(conditionalNode, inputs) =>
+        // New output ports provided by the conditional node:
+        val newConditionalOutputPorts = (conditionalNode.outputPorts map { p => p.name -> p }).toMap
+        // New output ports provided by the newly-created GraphInputNodes:
+        val newGraphNodeInputProvidedOutputPorts = (inputs map { i => i.name -> i.singleOutputPort }).toMap
 
-        FoldState(acc.nodes + scatterNode ++ inputs, acc.availableInputs ++ newScatterOutputPorts ++ newInputOutputPorts)
+        FoldState(acc.nodes + conditionalNode ++ inputs, acc.availableInputs ++ newConditionalOutputPorts ++ newGraphNodeInputProvidedOutputPorts)
       }
```
Comment (mcovarr) [inline wom/src/main/scala/wdl4s/wdl/WdlGraphNode.scala]: lotta duplication between the scatter and conditional code

File: wom/src/test/scala/wdl4s/wdl/wom/WdlConditionalWomSpec.scala [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+import wdl4s.wdl.{WdlNamespace, WdlNamespaceWithWorkflow}
+import wdl4s.wom.graph._
+import wdl4s.wom.graph.GraphNodePort.ConditionalOutputPort
+
+class WdlConditionalWomSpec extends FlatSpec with Matchers {
+
+  behavior of "WdlNamespaces with if blocks"
+
+  it should "convert inputs from conditional tasks into workflow inputs, and conditional task outputs into workflow outputs" in {
+    val scatterTest =
```
Comment (mcovarr) [inline wom/src/test/scala/wdl4s/wdl/wom/WdlConditionalWomSpec.scala]: such copy

File: wom/src/test/scala/wdl4s/wdl/wom/WdlConditionalWomSpec.scala [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+      """task foo {
+        |  Int i
+        |  command { ... }
+        |  output {
+        |    String out = i
+        |  }
+        |}
+        |
+        |workflow scatter_test {
```
Comment (mcovarr) [inline wom/src/test/scala/wdl4s/wdl/wom/WdlConditionalWomSpec.scala]: so paste

File: wom/src/test/scala/wdl4s/wdl/wom/WdlConditionalWomSpec.scala [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+        |  Boolean b
+        |  if (b) {
+        |    call foo
+        |  }
+        |}""".stripMargin
+
+    val namespace = WdlNamespace.loadUsingSource(scatterTest, None, None).get.asInstanceOf[WdlNamespaceWithWorkflow]
+    import lenthall.validation.ErrorOr.ShortCircuitingFlatMap
+    val scatterTestGraph = namespace.womExecutable.flatMap(_.graph)
```
Comment (mcovarr) [inline wom/src/test/scala/wdl4s/wdl/wom/WdlConditionalWomSpec.scala]: much mistak

File: wom/src/test/scala/wdl4s/wdl/wom/WdlConditionalWomSpec.scala [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+
+    scatterTestGraph match {
+      case Valid(g) => validateGraph(g)
+      case Invalid(errors) => fail(s"Unable to build wom version of scatter foo from WDL: ${errors.toList.mkString("\n", "\n", "\n")}")
```
Comment (mcovarr) [inline wom/src/test/scala/wdl4s/wdl/wom/WdlConditionalWomSpec.scala]: rong words

File: wom/src/test/scala/wdl4s/wdl/wom/WdlConditionalWomSpec.scala [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+    }
+
+    def validateGraph(workflowGraph: Graph) = {
+
+      case class OuterGraphValidations(conditionalNode: ConditionalNode, xs_inputNode: GraphInputNode)
```
Comment (mcovarr) [inline wom/src/test/scala/wdl4s/wdl/wom/WdlConditionalWomSpec.scala]: second param sketchy ah

File: wom/src/test/scala/wdl4s/wdl/wom/WdlConditionalWomSpec.scala [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+
+        workflowGraph.nodes should be(Set(conditionalNode, foo_i_inputNode, b_inputNode, foo_out_output))
+        OuterGraphValidations(conditionalNode, foo_i_inputNode)
+      }
+
+      case class InnerGraphValidations(foo_out_innerOutput: GraphOutputNode)
+      def validateInnerGraph(validatedOuterGraph: OuterGraphValidations): InnerGraphValidations = {
+        val foo_i_innerInput = validatedOuterGraph.conditionalNode.innerGraph.nodes.collectFirst {
+          case gin: GraphInputNode if gin.name == "foo.i" => gin
+        }.getOrElse(fail("Scatter inner graph did not contain a GraphInputNode 'foo.i'"))
```
Comment (mcovarr) [inline wom/src/test/scala/wdl4s/wdl/wom/WdlConditionalWomSpec.scala]: o noes scatters everywhere

**Review by mcovarr — APPROVED** [review]
I did get a sense that there might be some code duplication that could be factored out.

File: wom/src/main/scala/wdl4s/wom/graph/ConditionalNode.scala [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+
+    val conditionalExpressionValidation: ErrorOr[InstantiatedExpression] = condition.instantiateExpression(graphNodeSetter)
+
+    val linkedInputPortsAndGraphInputNodes: Set[LinkedInputPort] = innerGraph.nodes.inputDefinitions.map(inputPortLinker)
+    val linkedInputPorts = linkedInputPortsAndGraphInputNodes.map(_.newInputPort)
```
Comment (danbills) [inline wom/src/main/scala/wdl4s/wom/graph/ConditionalNode.scala]: I'm having trouble following these 3 lines of code, perhaps being explicit about types and maybe an overview of the goal?

**Review by danbills — APPROVED** [review]

File: wom/src/main/scala/wdl4s/wom/graph/ConditionalNode.scala (hunk 4 above)
Comment (cjllanwarne) [reply wom/src/main/scala/wdl4s/wom/graph/ConditionalNode.scala]: The first is a WDL specific error, this one is checking at the WOM level so is unable to be so specific.

**Pull request merged**